File: educational-core/src/com/jetbrains/edu/learning/checkio/checker/CheckiOMissionCheck.java
package com.jetbrains.edu.learning.checkio.checker;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Computable;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.learning.EduUtils;
import com.jetbrains.edu.learning.checker.CheckResult;
import com.jetbrains.edu.learning.checkio.api.exceptions.NetworkException;
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector;
import com.jetbrains.edu.learning.checkio.courseFormat.CheckiOMission;
import com.jetbrains.edu.learning.checkio.notifications.errors.handlers.CheckiOErrorHandler;
import com.jetbrains.edu.learning.checkio.utils.CheckiONames;
import com.jetbrains.edu.learning.courseFormat.CheckStatus;
import com.jetbrains.edu.learning.courseFormat.TaskFile;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.ui.taskDescription.BrowserWindow;
import javafx.application.Platform;
import javafx.concurrent.Worker;
import javafx.embed.swing.JFXPanel;
import netscape.javascript.JSObject;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.html.HTMLFormElement;
import org.w3c.dom.html.HTMLInputElement;
import org.w3c.dom.html.HTMLTextAreaElement;

import java.io.IOException;
import java.util.concurrent.Callable;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class CheckiOMissionCheck implements Callable<CheckResult> {
  private final Project myProject;
  private final Task myTask;

  private final BrowserWindow myBrowserWindow;
  private final CheckiOTestResultHandler myResultHandler;

  @Nullable private CheckResult myCheckResult;
  private final CountDownLatch myLatch = new CountDownLatch(1);
  private final CheckiOOAuthConnector myOAuthConnector;
  private final String myInterpreterName;
  private final String myTestFormTargetUrl;

  protected CheckiOMissionCheck(
    @NotNull Task task,
    @NotNull Project project,
    @NotNull CheckiOOAuthConnector oAuthConnector,
    @NotNull String interpreterName,
    @NotNull String testFormTargetUrl
  ) {
    myProject = project;
    myTask = task;
    myOAuthConnector = oAuthConnector;
    myInterpreterName = interpreterName;
    myTestFormTargetUrl = testFormTargetUrl;

    myResultHandler = new CheckiOTestResultHandler();
    myBrowserWindow = new BrowserWindow(myProject, false, false);
  }

  @NotNull
  @Override
  public CheckResult call() {
    try {
      final String accessToken = myOAuthConnector.getAccessToken();
      final String taskId = String.valueOf(myTask.getId());
      final String code = getCodeFromTask();

      return doCheck(accessToken, taskId, code);
    } catch (InterruptedException e) {
      return new CheckResult(CheckStatus.Unchecked, "Checking was cancelled");
    } catch (Exception e) {
      new CheckiOErrorHandler(
        "Failed to check the task",
        myOAuthConnector
      ).handle(e);
      return CheckResult.FAILED_TO_CHECK;
    }
  }

  private String getCodeFromTask() throws IOException {
    final TaskFile taskFile = ((CheckiOMission) myTask).getTaskFile();
    final VirtualFile missionDir = myTask.getDir(myProject);
    if (missionDir == null) {
      throw new IOException("Directory is not found for mission: " + myTask.getStepId() + ", " + myTask.getName());
    }

    final VirtualFile virtualFile = EduUtils.findTaskFileInDir(taskFile, missionDir);
    if (virtualFile == null) {
      throw new IOException("Virtual file is not found for mission: " + myTask.getStepId() + ", " + myTask.getName());
    }

    final Document document = ApplicationManager.getApplication().runReadAction((Computable<Document>) () ->
      FileDocumentManager.getInstance().getDocument(virtualFile)
    );

    if (document == null) {
      throw new IOException("Document isn't provided for VirtualFile: " + virtualFile.getName());
    }

    return document.getText();
  }

  @NotNull
  private CheckResult doCheck(@NotNull String accessToken, @NotNull String taskId, @NotNull String code)
    throws InterruptedException, NetworkException {

    Platform.runLater(() -> {
      setTestFormLoadedListener(accessToken, taskId, code);
      setCheckDoneListener();
      loadTestForm();
    });

    boolean timeoutExceeded = !myLatch.await(30L, TimeUnit.SECONDS);
    if (timeoutExceeded) {
      return new CheckResult(CheckStatus.Unchecked, "Checking took too much time");
    }

    if (myCheckResult == CheckResult.CONNECTION_FAILED) {
      throw new NetworkException();
    }

    //noinspection ConstantConditions cannot be null because of handler implementation
    return myCheckResult;
  }

  private void loadTestForm() {
    final String formUrl = getClass().getResource(CheckiONames.CHECKIO_TEST_FORM_URL).toExternalForm();
    myBrowserWindow.getEngine().load(formUrl);
  }

  private void setCheckDoneListener() {
    final Ref<Boolean> visited = new Ref<>(Boolean.FALSE);

    myBrowserWindow.getEngine().getLoadWorker().stateProperty().addListener((observable, oldState, newState) -> {
      if (newState == Worker.State.FAILED) {
        setConnectionError();
        return;
      }

      if (myBrowserWindow.getEngine().getLocation().contains(CheckiONames.CHECKIO_URL) && newState == Worker.State.SUCCEEDED && !visited.get()) {
        visited.set(Boolean.TRUE);

        final JSObject windowObject = (JSObject)myBrowserWindow.getEngine().executeScript("window");
        windowObject.setMember("javaHandler", myResultHandler);

        myBrowserWindow.getEngine().executeScript(
          "function handleEvent(e) {\n" +
          "\twindow.javaHandler.handleTestEvent(e.detail.success)\n" +
          "}\n" +
          "window.addEventListener(\"checkio:checkDone\", handleEvent, false)"
        );
      }
    });
  }

  private void setTestFormLoadedListener(@NotNull String accessToken,
                                         @NotNull String taskId,
                                         @NotNull String code) {
    myBrowserWindow.getEngine().getLoadWorker().stateProperty().addListener(((observable, oldState, newState) -> {
      if (newState == Worker.State.FAILED) {
        setConnectionError();
        return;
      }

      if (newState == Worker.State.SUCCEEDED && myBrowserWindow.getEngine().getLocation().contains("checkioTestForm.html")) {
        final org.w3c.dom.Document documentWithForm = myBrowserWindow.getEngine().getDocument();
        ((HTMLInputElement)documentWithForm.getElementById("access-token")).setValue(accessToken);
        ((HTMLInputElement)documentWithForm.getElementById("task-id")).setValue(taskId);
        ((HTMLInputElement)documentWithForm.getElementById("interpreter")).setValue(myInterpreterName);
        ((HTMLTextAreaElement)documentWithForm.getElementById("code")).setValue(code);

        final HTMLFormElement testForm = (HTMLFormElement) documentWithForm.getElementById("test-form");
        testForm.setAction(myTestFormTargetUrl);
        testForm.submit();
      }
    }));
  }

  private void setConnectionError() {
    myCheckResult = CheckResult.CONNECTION_FAILED;
    myLatch.countDown();
  }

  public JFXPanel getBrowserPanel() {
    return myBrowserWindow.getPanel();
  }

  public class CheckiOTestResultHandler {
    @SuppressWarnings("unused") // used in JS code
    public void handleTestEvent(int result) {
      myCheckResult = result == 1 ?
                      new CheckResult(CheckStatus.Solved, "All tests passed") :
                      new CheckResult(CheckStatus.Failed, "Tests failed");
      myLatch.countDown();
    }
  }
}

File: educational-core/src/com/jetbrains/edu/learning/checkio/checker/CheckiOMissionCheck.kt
package com.jetbrains.edu.learning.checkio.checker

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.invokeLater
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Computable
import com.intellij.openapi.util.Disposer
import com.intellij.ui.jcef.JBCefApp
import com.intellij.ui.jcef.JBCefBrowserBase
import com.intellij.ui.jcef.JBCefJSQuery
import com.intellij.ui.jcef.JCEFHtmlPanel
import com.jetbrains.edu.learning.EduUtils
import com.jetbrains.edu.learning.StudyTaskManager
import com.jetbrains.edu.learning.checker.CheckResult
import com.jetbrains.edu.learning.checkio.api.exceptions.NetworkException
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector
import com.jetbrains.edu.learning.checkio.courseFormat.CheckiOMission
import com.jetbrains.edu.learning.checkio.exceptions.CheckiOLoginRequiredException
import com.jetbrains.edu.learning.checkio.notifications.errors.CheckiOErrorReporter
import com.jetbrains.edu.learning.courseDir
import com.jetbrains.edu.learning.courseFormat.CheckStatus
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.courseGeneration.GeneratorUtils
import com.jetbrains.edu.learning.messages.EduCoreBundle
import org.cef.browser.CefBrowser
import org.cef.browser.CefFrame
import org.cef.handler.CefLoadHandler
import org.cef.handler.CefLoadHandlerAdapter
import org.jetbrains.annotations.NonNls
import java.io.IOException
import java.util.concurrent.Callable
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import javax.swing.JComponent

class CheckiOMissionCheck(private val project: Project,
                          private val task: Task,
                          private val oAuthConnector: CheckiOOAuthConnector,
                          @NonNls private val interpreterName: String,
                          @NonNls private val testFormTargetUrl: String
) : Callable<CheckResult> {
  private val jbCefBrowser = JCEFHtmlPanel(JBCefApp.getInstance().createClient(), null)
  private val jbCefJSQuery = JBCefJSQuery.create(jbCefBrowser as JBCefBrowserBase)
  private lateinit var checkResult: CheckResult
  private val latch = CountDownLatch(1)
  val panel: JComponent
    get() = jbCefBrowser.component

  init {
    jbCefBrowser.cefBrowser.createImmediately()
    jbCefBrowser.jbCefClient.addLoadHandler(TestFormLoadHandler(), jbCefBrowser.cefBrowser)

    jbCefJSQuery.addHandler { value ->
      val result = value.toIntOrNull() ?: return@addHandler null

      ApplicationManager.getApplication().executeOnPooledThread {
        setCheckResult(result)
      }
      null
    }
    // TODO: pass another disposable with shorter lifetime
    // for example, content manager of the corresponding tool window
    // otherwise, jcef browser objects can be leaked until the project is closed.
    // But it's better than nothing
    Disposer.register(StudyTaskManager.getInstance(project), jbCefBrowser)
  }

  @Throws(InterruptedException::class, NetworkException::class)
  override fun call(): CheckResult {
    return try {
      doCheck(collectResources())

      val timeoutExceeded: Boolean = !latch.await(30L, TimeUnit.SECONDS)
      if (timeoutExceeded) {
        return CheckResult(CheckStatus.Unchecked, EduCoreBundle.message("edu.check.took.too.much.time"))
      }

      if (checkResult === CheckResult.CONNECTION_FAILED) {
        throw NetworkException()
      }

      return checkResult
    }
    catch (e: CheckiOLoginRequiredException) {
      CheckiOErrorReporter(project, EduCoreBundle.message("label.login.required"), oAuthConnector).handle(e)
      CheckResult.LOGIN_NEEDED
    }
    catch (e: InterruptedException) {
      CheckResult(CheckStatus.Unchecked, EduCoreBundle.message("edu.check.was.cancelled"))
    }
    catch (e: Exception) {
      CheckiOErrorReporter(project, EduCoreBundle.message("notification.title.failed.to.check.task"), oAuthConnector).handle(e)
      CheckResult.failedToCheck
    }
    finally {
      latch.countDown()
    }
  }

  private fun doCheck(resources: Map<String, String>) {
    invokeLater {
      val html = getTestFormHtml(resources)
      jbCefBrowser.loadHTML(html)
    }
  }

  private fun getTestFormHtml(resources: Map<String, String>): String =
    GeneratorUtils.getInternalTemplateText(CHECKIO_TEST_FORM_TEMPLATE, resources)

  private fun setCheckResult(result: Int) {
    checkResult = when (result) {
      1 -> CheckResult(CheckStatus.Solved, EduCoreBundle.message("edu.check.all.tests.passed"))
      else -> CheckResult(CheckStatus.Failed, EduCoreBundle.message("edu.check.tests.failed"))
    }
    latch.countDown()
  }

  private fun collectResources(): Map<String, String> = mapOf(
    "testFormTargetUrl" to testFormTargetUrl,
    "accessToken" to oAuthConnector.getAccessToken(),
    "taskId" to task.id.toString(),
    "interpreterName" to interpreterName,
    "code" to getCodeFromTask()
  )

  private fun setConnectionError() {
    checkResult = CheckResult.CONNECTION_FAILED
    latch.countDown()
  }

  private fun getCodeFromTask(): String {
    val taskFile = (task as CheckiOMission).getTaskFile()
    val missionDir = task.getDir(project.courseDir)
                     ?: throw IOException("Directory is not found for mission: ${task.id}, ${task.name}")
    val virtualFile = EduUtils.findTaskFileInDir(taskFile, missionDir)
                      ?: throw IOException("Virtual file is not found for mission: ${task.id}, ${task.name}")

    val document = ApplicationManager.getApplication().runReadAction(
      Computable {
        FileDocumentManager.getInstance().getDocument(virtualFile)
      }
    ) ?: throw IOException("Document isn't provided for VirtualFile: ${virtualFile.name}")

    return document.text
  }

  private inner class TestFormLoadHandler : CefLoadHandlerAdapter() {
    override fun onLoadError(browser: CefBrowser?,
                             frame: CefFrame?,
                             errorCode: CefLoadHandler.ErrorCode?,
                             errorText: String?,
                             failedUrl: String?) {
      ApplicationManager.getApplication().executeOnPooledThread {
        setConnectionError()
      }
    }

    override fun onLoadEnd(browser: CefBrowser, frame: CefFrame?, httpStatusCode: Int) {
      // actually it means test form
      if (browser.url.contains("about:blank")) {
        browser.mainFrame.executeJavaScript(
          """
          form = document.getElementById('test-form');
          if (form) {
            form.hidden = true;
            form.submit();
          }
          """.trimIndent(), jbCefBrowser.cefBrowser.url, 0
        )
      }

      if (browser.url.contains("check-html-output")) {
        browser.mainFrame.executeJavaScript(
          """
          function handleEvent(e) {
            let value = e.detail.success;
            ${jbCefJSQuery.inject("value")}
          }
          window.addEventListener('checkio:checkDone', handleEvent, false);  
            
          document.documentElement.setAttribute("style", "background-color : #DEE7F6;")
          """.trimIndent(), jbCefBrowser.cefBrowser.url, 0
        )
      }
    }
  }

  companion object {
    @NonNls
    private const val CHECKIO_TEST_FORM_TEMPLATE = "checkioTestForm.html"
  }
}
